sympy/polys/rings.py

```diff
@@ -3005,3 +3179,300 @@ def sqf_list(f, all=False):
+            all_polys.extend(coeffs_i)
+
+            # Quick exit:
+            if any(len(c) == 1 for c in coeffs_i):
+                ring = polynomials[0].ring
+                domain = ring.domain
+                gcd = gcd_terms((all_polys + polynomials[i+1:]), ring, domain)
+                return [gcd], None
+
+def gcd_prs(polynomials):
+    """
+    Returns the greatest common divisor (GCD) of a set of polynomials.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> polynomials = x - y, x - y
+    >>> gcd_prs(polynomials)
+    x - y
+
+    """
+    ring = polynomials[0].ring
+    domain = ring.domain
+
+    if any(len(pi) == 1 for pi in polynomials):
+        return gcd_terms(polynomials, ring, domain)
+
+    polynomials, monom_gcd = monomial_extract(polynomials)
+
+    polynomials, common_symbols = _gcd_preprocess_polys(polynomials)
+
+    gcd = polynomials[0]
+    for pi in polynomials[1:]:
+        gcd = _gcd_prs(gcd, pi)
+        if gcd == domain.one:
+            break
+
+    if monom_gcd is not None:
+        gcd = gcd * monom_gcd
+
+    return gcd
+
+def _gcd_prs(p1, p2):
+    """
+    Returns the greatest common divisor (GCD) of two polynomials using the
+    Polynomial Resultant Sequences (PRS) method.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import _gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = 4*x**2 + 8*x + 10
+    >>> g = 2*x**2 + 6*x + 10
+    >>> _gcd_prs(f, g)
+    2
+
+    """
+    x = p1.main_variable()
+
+    c1, pp1 = cont_prim(p1, x)
+    c2, pp2 = cont_prim(p2, x)
+
+    h = pp1.subresultants(pp2, x)[-1]
+    c = gcd_prs([c1, c2])
+
+    ring = p1.ring
+    domain = ring.to_domain()
+
+
+    if domain.canonical_unit(h.coeff_wrt(x, h.degree(x))) or domain.is_Field:
+        h = -h
+
+    _, h = cont_prim(h, x)
+
+    h = h * c
+
+    return h
+
+def _gcd_(f, g):
+    """Helper function for ``_gcd_prs``."""
+
+    ring = f.ring
+    domain = ring.domain
+
+    if not domain.is_Exact:
+        try:
+            exact = domain.get_exact()
+        except DomainError:
+            return domain.one, f, g
+
+        f = domain.convert_from(f, exact)
+
+        g = domain.convert_from(g, exact)
+
+        h, cff, cfg  = _gcd_(f, g)
+
+        h = exact.convert_from(h, domain)
+        cff = exact.convert_from(cff, domain)
+        cfg = exact.convert_from(cfg, domain)
+
+        return h, cff, cfg
+
+    elif domain.is_Field:
+        if domain.is_QQ and query('USE_HEU_GCD'):
+            try:
+                return f._gcd_QQ(g)
+            except HeuristicGCDFailed:
+                pass
+
+        return f._gcd_I(g)
+
+    else:
+        if domain.is_ZZ and query('USE_HEU_GCD'):
+            try:
+                return f._gcd_ZZ(g)
+            except HeuristicGCDFailed:
+                pass
+
+        return f._gcd_I(g)
```

Comment:
Please check the implementation of this function, it is not working as expected

Reply:
The line ` f = domain.convert_from(f, exact)` supposes that `f` is an element of `exact` that should be converted to `domain`. This is wrong in two ways:

Firstly `exact` and `domain` are the wrong way around: we are trying to go from `domain` to `exact` not from `exact` to `domain`.

Secondly `f` is a polynomial and `exact` and `domain` refer to the domain of the polynomial like `QQ` and `RR` but `f` is not an element of either domain but rather an element of a polynomial ring over that domain like `RR[x,y]` or something.

The code should be something like:
```python
In [1]: old_ring = RR[x,y].ring

In [2]: old_ring
Out[2]: Polynomial ring in x, y over RR with lex order

In [3]: p = old_ring(x + 2.0*y)

In [4]: p
Out[4]: x + 2.0*y

In [5]: p.ring
Out[5]: Polynomial ring in x, y over RR with lex order

In [6]: old_domain = old_ring.domain

In [7]: old_domain
Out[7]: RR

In [8]: new_domain = old_domain.get_exact()

In [9]: new_domain
Out[9]: QQ

In [13]: new_ring = new_domain[old_ring.symbols].ring

In [14]: new_ring
Out[14]: Polynomial ring in x, y over QQ with lex order

In [15]: p
Out[15]: x + 2.0*y

In [16]: p.set_ring(new_ring)
Out[16]: x + 2*y
```